File: android/app/src/main/java/com/saratoga/Asr.kt
package com.saratoga

import com.cactus.cactusDestroy
import com.cactus.cactusInit
import com.cactus.cactusTranscribe
import org.json.JSONObject
import java.io.ByteArrayOutputStream
import java.io.DataOutputStream
import java.io.File

/** Moonshine-base ASR via Cactus. PCM bytes -> WAV file -> transcribe. */
class Asr(private val weightsDir: File) {
    private var handle: Long = 0L
    private val sampleRate = 16000

    fun load() {
        val path = File(weightsDir, "moonshine-base").absolutePath
        handle = cactusInit(path, null, false)
    }

    fun transcribe(pcm: ByteArray, tmpDir: File): String {
        val wav = File(tmpDir, "rec.wav")
        pcmToWav(pcm, wav)
        val raw = cactusTranscribe(handle, wav.absolutePath, null, null, null, null)
        val result = JSONObject(raw)
        return result.optString("response").trim()
    }

    fun close() {
        if (handle != 0L) { cactusDestroy(handle); handle = 0L }
    }

    private fun pcmToWav(pcm: ByteArray, out: File) {
        val byteRate = sampleRate * 1 * 16 / 8
        val totalDataLen = pcm.size + 36
        out.outputStream().use { fos ->
            val d = DataOutputStream(fos)
            val header = ByteArrayOutputStream().apply {
                write("RIFF".toByteArray())
                write(intLE(totalDataLen))
                write("WAVE".toByteArray())
                write("fmt ".toByteArray())
                write(intLE(16))          // fmt chunk size
                write(shortLE(1))         // PCM
                write(shortLE(1))         // mono
                write(intLE(sampleRate))
                write(intLE(byteRate))
                write(shortLE(2))         // block align
                write(shortLE(16))        // bits/sample
                write("data".toByteArray())
                write(intLE(pcm.size))
            }
            d.write(header.toByteArray())
            d.write(pcm)
        }
    }

    private fun intLE(v: Int) = byteArrayOf(
        (v and 0xff).toByte(), (v shr 8 and 0xff).toByte(),
        (v shr 16 and 0xff).toByte(), (v shr 24 and 0xff).toByte()
    )
    private fun shortLE(v: Int) = byteArrayOf((v and 0xff).toByte(), (v shr 8 and 0xff).toByte())
}

File: ios/SaratogaApp/Asr.swift
import AVFoundation
import Foundation
import cactus

/// Moonshine-base ASR via Cactus Apple SDK. PCM -> WAV -> transcribe.
final class Asr {
    private var model: CactusModelT?
    private let sampleRate = 16_000
    private let whisperPrompt = "<|startoftranscript|><|en|><|transcribe|><|notimestamps|>"

    func load(from weightsDir: URL) throws {
        let dir = weightsDir.appendingPathComponent("moonshine-base")
        Saratoga.unprotectTree(dir)
        model = try cactusInit(dir.path, nil, false)
    }

    func transcribe(pcm: Data, tmpDir: URL) throws -> String {
        guard let model = model else { return "" }
        let wav = tmpDir.appendingPathComponent("rec.wav")
        try writeWav(pcm: pcm, to: wav)
        return try transcribeFile(at: wav)
    }

    func transcribeFile(at url: URL) throws -> String {
        guard let model = model else { return "" }
        let pcm = try Asr.extractPCM16(from: url)
        NSLog("[asr] extracted pcm bytes=\(pcm.count) (~\(pcm.count / 32000)s)")
        let raw = try cactusTranscribe(model, nil, nil, nil, nil, pcm)
        NSLog("[asr] RAW JSON: \(raw)")
        guard let data = raw.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return "" }
        var text = (obj["response"] as? String) ?? ""
        if text.isEmpty, let segs = obj["segments"] as? [[String: Any]] {
            text = segs.compactMap { $0["text"] as? String }.joined(separator: " ")
        }
        NSLog("[asr] parsed text=\(text.prefix(120))")
        return text.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private static func extractPCM16(from url: URL) throws -> Data {
        let file = try AVAudioFile(forReading: url)
        let format = file.processingFormat
        let frameCount = AVAudioFrameCount(file.length)
        guard let source = AVAudioPCMBuffer(pcmFormat: format,
                                             frameCapacity: frameCount) else {
            throw NSError(domain: "Asr", code: -10,
                          userInfo: [NSLocalizedDescriptionKey: "buffer alloc failed"])
        }
        try file.read(into: source)

        guard let targetFormat = AVAudioFormat(
            commonFormat: .pcmFormatInt16,
            sampleRate: 16000,
            channels: 1,
            interleaved: true
        ) else {
            throw NSError(domain: "Asr", code: -11,
                          userInfo: [NSLocalizedDescriptionKey: "target format nil"])
        }
        if format.sampleRate == 16000 && format.channelCount == 1
            && format.commonFormat == .pcmFormatInt16 {
            // already target — grab bytes directly
            if let ch = source.int16ChannelData {
                let count = Int(source.frameLength) * MemoryLayout<Int16>.size
                return Data(bytes: ch[0], count: count)
            }
        }
        guard let converter = AVAudioConverter(from: format, to: targetFormat) else {
            throw NSError(domain: "Asr", code: -12,
                          userInfo: [NSLocalizedDescriptionKey: "converter nil"])
        }
        let outCap = AVAudioFrameCount(Double(source.frameLength) *
                                        (16000.0 / format.sampleRate)) + 1024
        guard let out = AVAudioPCMBuffer(pcmFormat: targetFormat, frameCapacity: outCap) else {
            throw NSError(domain: "Asr", code: -13,
                          userInfo: [NSLocalizedDescriptionKey: "out buffer nil"])
        }
        var err: NSError?
        var provided = false
        _ = converter.convert(to: out, error: &err) { _, status in
            if provided {
                status.pointee = .endOfStream
                return nil
            }
            status.pointee = .haveData
            provided = true
            return source
        }
        if let err = err { throw err }
        guard let ch = out.int16ChannelData else { return Data() }
        let count = Int(out.frameLength) * MemoryLayout<Int16>.size
        return Data(bytes: ch[0], count: count)
    }

    func close() {
        if let m = model { cactusDestroy(m); model = nil }
    }

    private func writeWav(pcm: Data, to url: URL) throws {
        var header = Data()
        let dataLen = UInt32(pcm.count)
        let byteRate = UInt32(sampleRate * 1 * 16 / 8)
        header.append("RIFF".data(using: .ascii)!)
        header.append(u32(UInt32(36 + Int(dataLen))))
        header.append("WAVE".data(using: .ascii)!)
        header.append("fmt ".data(using: .ascii)!)
        header.append(u32(16))
        header.append(u16(1))              // PCM
        header.append(u16(1))              // mono
        header.append(u32(UInt32(sampleRate)))
        header.append(u32(byteRate))
        header.append(u16(2))              // block align
        header.append(u16(16))             // bits/sample
        header.append("data".data(using: .ascii)!)
        header.append(u32(dataLen))
        var out = Data()
        out.append(header); out.append(pcm)
        try out.write(to: url)
    }

    private func u16(_ v: UInt16) -> Data {
        var x = v.littleEndian
        return Data(bytes: &x, count: 2)
    }
    private func u32(_ v: UInt32) -> Data {
        var x = v.littleEndian
        return Data(bytes: &x, count: 4)
    }
}
